indexer.go

```diff
@@ -0,0 +1,863 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/ipfs/go-cid"
+	"io"
+	"strings"
+
+	datastore "github.com/ipfs/go-datastore"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+
+	car2 "github.com/ipld/go-car/v2"
+	selectorparse "github.com/ipld/go-ipld-prime/traversal/selector/parse"
+
+	ipld "github.com/ipld/go-ipld-prime"
+
+	dagcbor "github.com/ipld/go-ipld-prime/codec/cbor"
+	dagjson "github.com/ipld/go-ipld-prime/codec/json"
+
+	datamodel "github.com/ipld/go-ipld-prime/datamodel"
+	qp "github.com/ipld/go-ipld-prime/fluent/qp"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	basicnode "github.com/ipld/go-ipld-prime/node/basicnode"
+
+	tree "github.com/kenlabs/go-ipld-prolly-trees/pkg/tree"
+)
+
+type Database struct {
+	blockStore  *blockstore.Blockstore
+	nodeStore   *tree.BlockNodeStore
+	rootCid     cid.Cid
+	tree        *tree.ProllyTree
+	collections map[string]*Collection
+}
+
+type Collection struct {
+	db         *Database
+	name       string
+	primaryKey []string
+}
+
+type Index struct {
+	collection *Collection
+	fields     []string
+}
+
+type Record struct {
+	Id   []byte
+	Data ipld.Node
+}
+
+type Query struct {
+	Equal map[string]ipld.Node
+}
+
+var NULL_BYTE = []byte("\x00")
+var FULL_BYTE = []byte("\xFF")
+var DATA_PREFIX = []byte("\x00d")
+var INDEX_PREFIX = []byte("\x00i")
+var DB_METADATA_KEY = []byte("\xFF\x00")
+var DB_VERSION = int64(1)
+var INDEX_VERSION_1 = int64(1)
+var VERSION_KEY = "version"
+
+func NewDatabaseFromBlockStore(ctx context.Context, blockStore blockstore.Blockstore) (*Database, error) {
+	nodeStore, err := tree.NewBlockNodeStore(blockStore, &tree.StoreConfig{CacheSize: 1 << 10})
+
+	if err != nil {
```

Comment:
Same as you have written, but without the newline (line 70 in this case). This is very minor of course, I just mention it because it looks unfamiliar and catches my attention when reading.